worker/instancepoller/worker.go

```diff
@@ -85,57 +180,319 @@ func (u *updaterWorker) Wait() error {
 	return u.catacomb.Wait()
 }
 
-func (u *updaterWorker) loop() (err error) {
-	u.aggregator, err = newAggregator(
-		aggregatorConfig{
-			Clock:         u.config.Clock,
-			Delay:         u.config.Delay,
-			Environ:       u.config.Environ,
-			CredentialAPI: u.config.CredentialAPI,
-		},
-	)
+func (u *updaterWorker) loop() error {
+	watcher, err := u.config.Facade.WatchModelMachines()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if err := u.catacomb.Add(u.aggregator); err != nil {
+	if err := u.catacomb.Add(watcher); err != nil {
 		return errors.Trace(err)
 	}
-	watcher, err := u.config.Facade.WatchModelMachines()
+
+	shortPollTimer := u.config.Clock.NewTimer(ShortPoll)
+	longPollTimer := u.config.Clock.NewTimer(LongPoll)
+	defer func() {
+		_ = shortPollTimer.Stop()
+		_ = longPollTimer.Stop()
+	}()
+
+	for {
+		select {
+		case <-u.catacomb.Dying():
+			return u.catacomb.ErrDying()
+		case ids, ok := <-watcher.Changes():
+			if !ok {
+				return errors.New("machines watcher closed")
+			}
+
+			for i := range ids {
+				tag := names.NewMachineTag(ids[i])
+				if err := u.queueMachineForPolling(tag); err != nil {
+					return err
+				}
+			}
+		case <-shortPollTimer.Chan():
```

Comment:
TBH, I thought of implementing something like this but wasn't sure that it was needed so I decided to push the code as-is and wait for feedback to avoid premature optimizations.

As you mentioned iterating 1k (or 10k) items and doing a quick int64 comparison every second shouldn't really be a big issue. That being said, I can easily implement the min-time-before-waking up feature so I will give that a go (although it will make tests a bit more complicated). Instead of adding some fudge quantity we can just round up the min-time to the next poll interval.

There is also another micro-optimization that we can implement if we want: we can allocate the list with the group entries to be polled once when the worker is created and reset its size to 0 (and retain the capacity) when we iterate the poll group to populate the list.